src/runtime/rpc/group_host_port.h

```diff
@@ -0,0 +1,221 @@
+    }
+    void set_leader(host_port hp);
+    bool remove(host_port hp) WARN_UNUSED_RESULT;
+    bool contains(host_port hp) const WARN_UNUSED_RESULT;
+    int count();
+
+    const std::vector<host_port> &members() const { return _members; }
+    host_port random_member() const
+    {
+        alr_t l(_lock);
+        return _members.empty() ? host_port::s_invalid_host_port
+                                : _members[rand::next_u32(0, (uint32_t)_members.size() - 1)];
+    }
+    host_port next(host_port current) const;
+    host_port leader() const
+    {
+        alr_t l(_lock);
+        return _leader_index >= 0 ? _members[_leader_index] : host_port::s_invalid_host_port;
+    }
+    void leader_forward();
+    host_port possible_leader();
+    bool is_update_leader_automatically() const { return _update_leader_automatically; }
+    void set_update_leader_automatically(bool value) { _update_leader_automatically = value; }
+    const char *name() const { return _name.c_str(); }
+
+private:
+    typedef std::vector<host_port> members_t;
+    typedef utils::auto_read_lock alr_t;
+    typedef utils::auto_write_lock alw_t;
+
+    mutable utils::rw_lock_nr _lock;
+    members_t _members;
+    int _leader_index;
+    bool _update_leader_automatically;
+    std::string _name;
+};
+
+// ------------------ inline implementation --------------------
+
+inline rpc_group_host_port::rpc_group_host_port(const char *name)
+{
+    _name = name;
+    _leader_index = -1;
+    _update_leader_automatically = true;
+}
+
+inline rpc_group_host_port::rpc_group_host_port(const rpc_group_host_port &other)
+{
+    _name = other._name;
+    _leader_index = other._leader_index;
+    _update_leader_automatically = other._update_leader_automatically;
+    _members = other._members;
+}
+
+inline rpc_group_host_port &rpc_group_host_port::operator=(const rpc_group_host_port &other)
+{
+    if (this == &other) {
+        return *this;
+    }
+    _name = other._name;
+    _leader_index = other._leader_index;
+    _update_leader_automatically = other._update_leader_automatically;
+    _members = other._members;
+    return *this;
+}
+
+inline bool rpc_group_host_port::add(host_port hp)
+{
+    CHECK_EQ_MSG(hp.type(), HOST_TYPE_IPV4, "rpc group host_port member must be ipv4");
+
+    alw_t l(_lock);
+    if (_members.end() == std::find(_members.begin(), _members.end(), hp)) {
+        _members.push_back(hp);
+        return true;
+    } else {
+        return false;
+    }
+}
+
+inline void rpc_group_host_port::leader_forward()
+{
+    alw_t l(_lock);
+    if (_members.empty()) {
+        return;
+    }
+    _leader_index = (_leader_index + 1) % _members.size();
+}
+
+inline void rpc_group_host_port::set_leader(host_port hp)
+{
+    alw_t l(_lock);
+    if (hp.is_invalid()) {
+        _leader_index = -1;
+        return;
+    }
+
+    CHECK_EQ_MSG(hp.type(), HOST_TYPE_IPV4, "rpc group host_port member must be ipv4");
+    for (int i = 0; i < (int)_members.size(); i++) {
```

Comment:
```suggestion
    for (int i = 0; i < static_cast<int>(_members.size()); i++) {
```
Some other places are the same.